aemcmc/gibbs.py

```diff
@@ -364,6 +430,129 @@ def nbinom_horseshoe_step(
+    y: TensorVariable
+        The observed count data.
+    n_samples: TensorVariable
+        A tensor describing the number of posterior samples to generate.
+
+    Returns
+    -------
+    (outputs, updates): tuple
+        A symbolic description of the sampling result to be used to
+        compile a sampling function.
+
+    Notes
+    -----
+    The ``z`` expression in section 2.2 of [1] seems to
+    omit division by the Polya-Gamma auxilliary variables whereas [2] and [3]
+    explicitely include it. We found that including the division results in
+    accurate posterior samples for the regression coefficients. It is also
+    worth noting that the :math:`\sigma^2` parameter is not sampled directly
+    in the negative binomial regression problem and thus set to 1 [2].
+
+    References
+    ----------
+    ..[1] Makalic, Enes & Schmidt, Daniel. (2015). A Simple Sampler for the
+          Horseshoe Estimator. 10.1109/LSP.2015.2503725.
+    ..[2] Makalic, Enes & Schmidt, Daniel. (2016). High-Dimensional Bayesian
+          Regularised Regression with the BayesReg Package.
+    ..[3] Neelon, Brian. (2019). Bayesian Zero-Inflated Negative Binomial
+          Regression Based on Pólya-Gamma Mixtures. Bayesian Anal.
+          2019 September ; 14(3): 829–855. doi:10.1214/18-ba1132.
+
+    """
+
+    def nbinom_horseshoe_step(
+        beta: TensorVariable,
+        lmbda: TensorVariable,
+        tau: TensorVariable,
+        r: TensorVariable,
+        y: TensorVariable,
+        X: TensorVariable,
+        a_0: TensorVariable,
+    ) -> Tuple[TensorVariable, TensorVariable, TensorVariable, TensorVariable]:
+        """Complete one full update of the gibbs sampler and return the new state
+        of the posterior conditional parameters.
+
+        Parameters
+        ----------
+        beta: Tensorvariable
+            Coefficients (other than intercept) of the regression model.
+        lmbda
+            Inverse of the local shrinkage parameter of the horseshoe prior.
+        tau
+            Inverse of the global shrinkage parameters of the horseshoe prior.
+        y: TensorVariable
+            The observed count data.
+        X: TensorVariable
+            The covariate matrix.
+        r: TensorVariable
+            The "number of successes" parameter of the negative binomial disribution
+            used to model the data.
+        a_0 : TensorVariable
+            The shape parameter for r gamma prior
+
+        """
+        xb = X @ beta
+        w = srng.gen(polyagamma, y + r, xb)
+        z = 0.5 * (y - r) / w
+
+        lmbda_inv = 1.0 / lmbda
+        tau_inv = 1.0 / tau
+        beta_new = update_beta(srng, w, lmbda_inv * tau_inv, X, z)
+
+        lmbda_inv_new, tau_inv_new = horseshoe_step(
+            srng, beta_new, 1.0, lmbda_inv, tau_inv
+        )
+        eta = X @ beta_new
+        p = at.sigmoid(-eta)
+        h = srng.gamma(1e-2 + 1 * 1e-2, 1 / (1e-2 + r))
+        r_new = r_step(srng, r, p, a_0, h, y)
```

Comment:
This `srng.gamma` is a fixed (hyper)prior assumption that we can't have baked into this sampler.  Instead, we want to let the user define their own models/(hyper)priors&mdash;and corresponding sample steps.
```suggestion
        r_new = r_step(srng, r, p, a_0, h, y)
```